File: src/Suspension/Suspension.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <fstream>

// Definir M_PI si no está definido
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

int main() {
    // Parámetros
    double m = 193.0 / 2.0;
    double g = 9.81;
    double Lo = 0.7;  // Elongación inicial en metros
    double dt = 0.01; // Paso de tiempo
    double totalTime = 10.0; // Tiempo total de simulación en segundos
    int steps = static_cast<int>(totalTime / dt);
    double k = m * M_PI * M_PI;
    double b = 0.25 * 2.0 * std::sqrt(k * m);

    // Condiciones iniciales
    double x = 0; // Desplazamiento inicial en metros
    double x_dot = 0; // Velocidad inicial en m/s
    double x_ddot; // Aceleración inicial en m/s^2

    // Vectores para almacenar resultados
    std::vector<double> times;
    std::vector<double> positions;

    // Bucle de simulación
    for (int i = 0; i < steps; ++i) {
        // Calcula la aceleración (x_ddot)
        x_ddot = (m * g - (x + Lo) * k - x_dot * b) / m;

        // Actualiza la velocidad y posición usando el método de Euler
        x_dot += x_ddot * dt;
        x += x_dot * dt;

        // Guarda los resultados
        times.push_back(i * dt);
        positions.push_back(0.7 - x);
    }

    // Escribir los resultados en un archivo CSV
    std::ofstream file("C:/Users/1M72696/Documents/1.MOISEBIM/Programas/srcsimulacion_suspension.csv");
    file << "Tiempo (s),Desplazamiento (m)\n";
    for (size_t i = 0; i < times.size(); ++i) {
        file << times[i] << "," << positions[i] << "\n";
    }
    file.close();

    std::cout << "Simulación completada. Resultados guardados en 'simulacion_suspension.csv'.\n";

    return 0;
}

File: src/Suspension/Suspension.py
import math
import matplotlib.pyplot as plt

# Parámetros
m = 193.0 / 2.0
g = 9.81
Lo = 0.7  # Elongación inicial en metros
dt = 0.01  # Paso de tiempo
totalTime = 10.0  # Tiempo total de simulación en segundos
steps = int(totalTime / dt)
k = m * math.pi * math.pi
b = 0.25 * 2.0 * math.sqrt(k * m)

# Condiciones iniciales
x = 0  # Desplazamiento inicial en metros
x_dot = 0  # Velocidad inicial en m/s
x_ddot = 0 # Aceleración inicial en m/s^2

# Listas para almacenar resultados
times = []
positions = []

# Bucle de simulación
for i in range(steps):
    # Calcula la aceleración (x_ddot)
    x_ddot = (m * g - (x + Lo) * k - x_dot * b) / m

    # Actualiza la velocidad y posición usando el método de Euler
    x_dot += x_ddot * dt
    x += x_dot * dt

    # Guarda los resultados
    times.append(i * dt)
    positions.append(0.7-x)

    # Imprime los resultados (opcional)
    print(f"Tiempo: {i * dt:.2f} s, Desplazamiento: {x:.4f} m, Velocidad: {x_dot:.4f} m/s, Aceleración: {x_ddot:.4f} m/s^2")

# Graficar el desplazamiento en función del tiempo
plt.figure(figsize=(10, 6))
plt.plot(times, positions, label='x (m)')
plt.xlabel('Tiempo (s)')
plt.ylabel('Desplazamiento (m)')
plt.title('Desplazamiento en función del tiempo')
plt.ylim(0,1)
plt.legend()
plt.grid(True)
plt.show()
